### PR TITLE
Fix caching of boundaries

### LINKED ISSUE #20495: MDNorm... algorithms dropping some points
When I run a script like 

```
w=Load(Filename='/SNS/HYS/IPTS-18741/nexus/HYS_174983.nxs.h5')
SetUB(Workspace=w, a=10.8, b=10.8, c=10.8, u='0,0,1', v='1,0,0')
dgs,_=DgsReduction(SampleInputWorkspace=w, SampleInputMonitorWorkspace=w, SofPhiEIsDistribution=False)
md=ConvertToMD(InputWorkspace=dgs, QDimensions='Q3D', 
               Q3DFrames='HKL', QConversionScales='HKL')
a,b=MDNormDirectSC(InputWorkspace=md, 
                   AlignedDim0='[0,0,L],0,5,200', 
                   AlignedDim1='[0,K,0],-.8,.8,100', 
                   AlignedDim2='[H,0,0],-4,1,1', 
                   AlignedDim3='DeltaE,-1,1,1')

cov=CalculateCoverageDGS(InputWorkspace=dgs, IncidentEnergy=18, 
                         Dimension1='Q3', Dimension1Min=0, Dimension1Max=5, Dimension1Step=0.025, 
                         Dimension2='Q2', Dimension2Min=-.8, Dimension2Max=.8, Dimension2Step=0.016, 
                         Dimension3='Q1', Dimension3Min=-4, Dimension3Max=1, Dimension3Step=5,
                         Dimension4='DeltaE', Dimension4Min=-1, Dimension4Max=1, Dimension4Step=2)
```

I expect that where I see non-zero pixels in the coverage image to get non-zero contributions in the normalization (b workspace). As you can see in the images below, I have some missing contribution around (0,1)

b
cov

### PR DESCRIPTION
**Pull request by AndreiSavici**
When caching the intersection planes, we used number of bins. However we have number of bins+1 boundaries for MD boxes.

This issue affects MDNorm algorithms, when the number of bins in 1 direction is small (especially 1)

**To test:**
Run the script in the issue. The new normalization should look like
b

Fixes #20495

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/MDAlgorithms/src/MDNormDirectSC.cpp [hunk 1]
```diff
@@ -405,29 +405,29 @@ void MDNormDirectSC::cacheDimensionXValues() {
                                (PhysicalConstants::h * PhysicalConstants::h);
   if (!m_hIntegrated) {
     auto &hDim = *m_normWS->getDimension(m_hIdx);
-    m_hX.resize(hDim.getNBins());
+    m_hX.resize(hDim.getNBins()+1);
```
Comment (quantumsteve) [inline Framework/MDAlgorithms/src/MDNormDirectSC.cpp]: What do you think about adding another member function to `IMDDimension` that returns the `getNBins() + 1`? Perhaps `IMDDImension::getNPoints()`

**Review by quantumsteve — CHANGES_REQUESTED** [review]
Congratulations on finding what's causing the difference between the normalization and coverage workspaces. Off-by-one errors are tough. I ran the script in the issue and verified that the missing contribution around (0,1) is resolved.

I think getting the number of points (one more than the number of bins) is common enough to justify it's own member function. Would you consider adding it to `IMDDimension`?

File: Framework/MDAlgorithms/src/MDNormDirectSC.cpp (hunk 1 above)
Comment (AndreiSavici) [reply Framework/MDAlgorithms/src/MDNormDirectSC.cpp]: I will do it as a separate issue. Would a name like `getNBoundaries` be more accurate?

Comment (quantumsteve) [reply Framework/MDAlgorithms/src/MDNormDirectSC.cpp]: `getNBoundaries` sounds better to me. Adding this should be one line of code. There's no reason to move it to a later PR.

**Comment by AndreiSavici** [comment]
Created issue #20497 to refactor NBins+1

**Comment by AndreiSavici** [comment]
Not sure why the ISIS test fails. It does not use these algorithms

**Comment by AndreiSavici** [comment]
The test passes on my machine

**Comment by quantumsteve** [comment]
So long as this is updated to use `getNBoundaries()` in PR #20541, LGTM.

**Review by quantumsteve — APPROVED** [review]

**Comment by quantumsteve** [comment]
closing since this is a duplicate of #20541.

**Pull request closed without merging**